### PR TITLE
integration: Publish aggregate test report summary

### PR DESCRIPTION
**Pull request by mqasimsarfraz**
Currently, getting overview of test failures in the CI can be challenging. This change addresses that by extending the work we did for summaries and enhances that by storing them to a GitHub page. 

It took inspiration from the `ig-benchmarks` and commits the reports on each change. [Similarly](https://github.com/benchmark-action/github-action-benchmark/pull/216/files), uses `retry` to avoid running into multiple `git push` issue.

## Testing Done

Reports: https://inspektor-gadget.github.io/ig-test-reports/

### CODE REVIEW

**Comment by alban** [comment]
That looks great.

Will the web page display all previous tests or only  the last few ones? If it displays everything, that would be a lot of columns over time, and make it difficult to read.

**Review by mauriciovasquezbernal — COMMENTED** [review]
It looks really great. I don't have so many comments, besides Alban's comment above I don't have anything to add. I think we should we the last N runs, N == 100?, and that's it.

**Comment by mqasimsarfraz** [comment]
> If it displays everything, that would be a lot of columns over time, and make it difficult to read.

:100: 

> Will the web page display all previous tests or only the last few ones? 

I currently limit it to [last 10](https://github.com/inspektor-gadget/ig-test-reports/blob/main/index.html#L90) only. We can change this value in ig-test-reports later on as well.

**Comment by mauriciovasquezbernal** [comment]
> I currently limit it to [last 10](https://github.com/inspektor-gadget/ig-test-reports/blob/main/index.html#L90) only. We can change this value in ig-test-reports later on as well.

Should we limit the size of the json somehow? I suppose it'll become to big and difficult to handle very soon.

**Comment by mqasimsarfraz** [comment]
> > I currently limit it to [last 10](https://github.com/inspektor-gadget/ig-test-reports/blob/main/index.html#L90) only. We can change this value in ig-test-reports later on as well.
> 
> Should we limit the size of the json somehow? I suppose it'll become to big and difficult to handle very soon.

Great catch! I adjusted to only store last [100 results](https://github.com/inspektor-gadget/inspektor-gadget/pull/2453/files#diff-bda633e49a0d242b6842493c3896d4f766f361166fa4c08e1c2e01d827544cadR1595). I believe we should be fine with that but if that will be too much we can decrease this in future.

File: .github/workflows/inspektor-gadget.yml
Comment (mauriciovasquezbernal) [inline .github/workflows/inspektor-gadget.yml:1562]: Perhaps move this to something under https://github.com/inspektor-gadget/inspektor-gadget/tree/main/tools to avoid adding more lines to this file?

**Comment by mauriciovasquezbernal** [comment]
> Also, with regard to the code itself, you make a parallel with `benchmarks` which has been really flaky recently...
> If possible, I would like to avoid adding other flaky tests in our CI as it just decreases the reliability over time.
> I understand this is indeed complicated to guess whether something will be flaky but the `retry` function does not really lead me to the path where I think this will not be flaky.

I think the logic introduced in the retry function is a good enough effort to avoid this job failing when other instances of it  run on parallel on different machines. Perhaps we can also use a concurrency group (https://docs.github.com/en/actions/using-jobs/using-concurrency) to avoid multiple instances of this job running on parallel and reduce the change of a collision?

**Review by mauriciovasquezbernal — COMMENTED** [review]
It's looking very good. What do you think about moving the script to a different file? I think that would help to edit it easier (highlight syntax and other stuff will help too). 

Also, I can approve it now if you want to merge as it is. I saw the result and it's very nice, we can always improve it later on if needed. It's up to you how much you want to improve it before merging.

**Comment by mqasimsarfraz** [comment]
> so use a concurrency group (https://docs.github.com/en/actions/using-jobs/using-concurrency) to avoid multiple instances of this job running on parallel and reduce the change of a collision?

I also thought about this but the reason I decided against it since it only allows single job in pending state?

> Any previously pending job or workflow in the concurrency group will be canceled. To also cancel any currently running job or workflow in the same concurrency group, specify cancel-in-progress: true.

File: .github/workflows/inspektor-gadget.yml
Comment (mqasimsarfraz) [reply .github/workflows/inspektor-gadget.yml:1562]: Good idea! I will do it!

**Comment by mauriciovasquezbernal** [comment]
So in that case we'll missing the entry for the skipped job in the results? 
I agree the current approach is better then, let's go for it.

File: .github/workflows/inspektor-gadget.yml
Comment (mqasimsarfraz) [reply .github/workflows/inspektor-gadget.yml]: Done.

**Review by mauriciovasquezbernal — APPROVED** [review]
LGTM. Thanks for working on this!

**Comment by mqasimsarfraz** [comment]
> > The issue with benchmarks was [#2038 (comment)](https://github.com/inspektor-gadget/inspektor-gadget/issues/2038#issuecomment-1910967021) so the idea was to see if we can handle the things in similar way to avoid issues in first place.
> 
> You made me doubt, but there is still an issue with benchmarks: https://github.com/inspektor-gadget/inspektor-gadget/actions/runs/7810164205/job/21303171942?pr=2460 Maybe this is a different one, but this will change that I deem this to be flaky as hell.

The error does look different but I see the point. But I don't think we should block this change on it. I can keep a close on eye on next few builds after we merge this change and would be happy to switch to other approaches (e.g concurrency group) if they are more stable yet with less coverage. Flaky CI is the last thing I want :smile:

**Comment by mqasimsarfraz** [comment]
I will merge this and keep a close eye on flakiness. Hopefully we will be able to identity failures faster!

**Pull request merged**
